src/ctapipe/tools/stats_calculation.py

```diff
@@ -0,0 +1,179 @@
+        if self.input_url == self.output_path:
+            raise ToolConfigurationError(
+                "Input and output files are same. Fix your configuration / cli arguments."
+            )
+
+        # Load the subarray description from the input file
+        subarray = SubarrayDescription.from_hdf(self.input_url)
+        # Initialization of the statistics calculator
+        self.stats_calculator = PixelStatisticsCalculator(
+            parent=self, subarray=subarray
+        )
+        # Read the input data with the 'TableLoader'
+        input_data = TableLoader(input_url=self.input_url)
+        # Get the telescope ids from the input data or use the allowed_tels configuration
+        tel_ids = subarray.tel_ids if self.allowed_tels is None else self.allowed_tels
+        # Read the whole dl1 images
+        self.dl1_tables = input_data.read_telescope_events_by_id(
+            telescopes=tel_ids,
+            dl1_images=True,
+            dl1_parameters=False,
+            dl1_muons=False,
+            dl2=False,
+            simulated=False,
+            true_images=False,
+            true_parameters=False,
+            instrument=False,
+            pointing=False,
+        )
+
+    def start(self):
+        # Iterate over the telescope ids and their corresponding dl1 tables
+        for tel_id, dl1_table in self.dl1_tables.items():
+            # Perform the first pass of the statistics calculation
+            aggregated_stats = self.stats_calculator.first_pass(
+                table=dl1_table,
+                tel_id=tel_id,
+                col_name=self.dl1a_column_name,
+            )
+            # Check if 'chunk_shift' is selected
+            if self.stats_calculator.chunk_shift is not None:
+                # Check if there are any faulty chunks to perform a second pass over the data
+                if np.any(~aggregated_stats["is_valid"].data):
+                    # Perform the second pass of the statistics calculation
+                    aggregated_stats_secondpass = self.stats_calculator.second_pass(
+                        table=dl1_table,
+                        valid_chunks=aggregated_stats["is_valid"].data,
+                        tel_id=tel_id,
+                        col_name=self.dl1a_column_name,
+                    )
+                    # Stack the statistic values from the first and second pass
+                    aggregated_stats = vstack(
+                        [aggregated_stats, aggregated_stats_secondpass]
+                    )
+                    # Sort the stacked aggregated statistic values by starting time
+                    aggregated_stats.sort(["time_start"])
+                else:
+                    self.log.info(
+                        "No faulty chunks found for telescope 'tel_id=%d'. Skipping second pass.",
+                        tel_id,
+                    )
+            # Write the aggregated statistics and their outlier mask to the output file
```

Comment:
I think logically output writing shall be in the `finish` function.

Reply:
That requires keeping all intermediate results for each telescope in memory until `finish`.

I'd write here, so only the data of a single telescope is in RAM at any given time.